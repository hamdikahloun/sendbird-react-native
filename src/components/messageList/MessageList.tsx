import React from 'react';
import {FlatList, SafeAreaView, StyleSheet, Text, View} from 'react-native';
import {UserMessage} from 'sendbird';
import {Message} from '../message';
import {LoadingState} from '../loadingState';

type Messages = UserMessage[];
type ChannelListProps = {
  connectedUserId: string;
  messages: Messages;
  loading: boolean;
  error: string;
  onEndReached?: () => void;
};

/**
 * MessageList - The message list component renders a list of messages.
 *
 * See:
 *
 * [BaseMessage](https://sendbird.github.io/core-sdk-javascript/module-model_baseMessage-BaseMessage.html,
 * @example
 *
 * <MessageList
 *  connectedUserId={connectedUserId}
 *  loading={loading}
 *  messages={messages}
 *  error={error}
 *  onEndReached={onEndReached}
 * />
 */
export const MessageList = ({
  messages,
  loading,
  error,
  connectedUserId,
  onEndReached,
}: ChannelListProps) => {
  const renderItem = ({item}: {item: UserMessage}) => (
    <Message
      isMyMessage={item.sender?.userId === connectedUserId}
      message={item}
    />
  );

  const ListEmptyComponent = () => {
    if (loading && !error?.length) {
      return <LoadingState />;
    }
    if (error) {
      return (
        <View style={styles.listEmptyComponent}>
          <Text>{error}</Text>
        </View>
      );
    }
    if (!loading && !error && !messages.length) {
      return (
        <View style={styles.listEmptyComponent}>
          <Text>Empty List</Text>
        </View>
      );
    }
    return <View />;
  };

  return (
    <SafeAreaView style={styles.container}>
      <FlatList
        inverted={messages.length > 0}
        style={styles.flatList}
        data={messages}
        renderItem={renderItem}
        keyExtractor={item => item.messageId.toString()}
        ListEmptyComponent={<ListEmptyComponent />}
        onEndReached={onEndReached}
        showsVerticalScrollIndicator={false}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f3F3F3',
  },
  flatList: {
    paddingHorizontal: 16,
  },
  listEmptyComponent: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  items: {
    flexWrap: 'wrap',
  },
});
