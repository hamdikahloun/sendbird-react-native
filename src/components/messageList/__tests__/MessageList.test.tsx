import React from 'react';
import {render} from '@/../testing';
import {MessageList} from '../MessageList';

describe('MessageList', () => {
  it('should render correctly', async () => {
    const screen = render(
      <MessageList
        messages={[]}
        loading={false}
        error={''}
        connectedUserId={''}
      />,
    ).toJSON();
    expect(screen).toMatchSnapshot();
  });
});
