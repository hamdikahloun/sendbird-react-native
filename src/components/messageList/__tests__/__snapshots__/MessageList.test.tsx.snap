// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList should render correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#f3F3F3",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={<ListEmptyComponent />}
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    invertStickyHeaders={false}
    inverted={false}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "paddingHorizontal": 16,
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text>
          Empty List
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
